internal/discovery/discovery.go

```diff
@@ -0,0 +1,223 @@
+				})
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			objSpec := obj.(*unstructured.Unstructured).Object["spec"].(map[string]interface{})
+			for _, version := range objSpec["versions"].([]interface{}) {
+				gotGVK := schema.GroupVersionKind{
+					Group:   objSpec["group"].(string),
+					Version: version.(map[string]interface{})["name"].(string),
+					Kind:    objSpec["names"].(map[string]interface{})["kind"].(string),
+				}
+				r.RemoveFromMap(gotGVK)
+				r.SafeWrite(func() {
+					r.WasUpdated = true
+				})
+			}
+		},
+	})
+	if err != nil {
+		return err
+	}
+	// Respect context cancellation.
+	go func() {
+		for range ctx.Done() {
+			klog.Infoln("Context cancelled, stopping discovery")
+			close(stopper)
+			return
+		}
+	}()
+	go informer.Run(stopper)
+	return nil
+}
+
+// ResolveGVK resolves the variable VKs to a GVK list, based on the current cache.
+func (r *CRDiscoverer) ResolveGVK(gvk schema.GroupVersionKind) (resolvedGVKs []schema.GroupVersionKind, err error) {
+	g := gvk.Group
+	v := gvk.Version
+	k := gvk.Kind
+	if g == "" || g == "*" {
+		return nil, fmt.Errorf("group is required in the defined GVK %v", gvk)
+	}
+	hasVersion := v != "" && v != "*"
+	hasKind := k != "" && k != "*"
+	// No need to resolve, return.
+	if hasVersion && hasKind {
+		return []schema.GroupVersionKind{
+			{Group: g, Version: v, Kind: k},
+		}, nil
+	}
+	if hasVersion && !hasKind {
+		kinds := r.Map[g][v]
+		for _, kind := range kinds {
+			resolvedGVKs = append(resolvedGVKs, schema.GroupVersionKind{Group: g, Version: v, Kind: kind})
+		}
+	}
+	if !hasVersion && hasKind {
+		versions := r.Map[g]
+		for version, kinds := range versions {
+			for _, kind := range kinds {
+				if kind == k {
+					resolvedGVKs = append(resolvedGVKs, schema.GroupVersionKind{Group: g, Version: version, Kind: k})
+				}
+			}
+		}
+	}
+	if !hasVersion && !hasKind {
+		versions := r.Map[g]
+		for version, kinds := range versions {
+			for _, kind := range kinds {
+				resolvedGVKs = append(resolvedGVKs, schema.GroupVersionKind{Group: g, Version: version, Kind: kind})
+			}
+		}
+	}
+	return
+}
+
+// PollForCacheUpdates polls the cache for updates and updates the stores accordingly.
+func (r *CRDiscoverer) PollForCacheUpdates(
+	ctx context.Context,
+	opts *options.Options,
+	storeBuilder *store.Builder,
+	m *metricshandler.MetricsHandler,
+	factoryGenerator func() ([]customresource.RegistryFactory, error),
+) {
+	// The interval at which we will check the cache for updates.
+	t := time.NewTicker(Interval)
+	// Track previous context to allow refreshing cache.
+	olderContext, olderCancel := context.WithCancel(ctx)
+	// Prevent context leak (kill the last metric handler instance).
+	defer olderCancel()
+	generateMetrics := func() {
+		// Get families for discovered factories.
+		customFactories, err := factoryGenerator()
+		if err != nil {
+			klog.Errorf("failed to update custom resource stores: %v", err)
+		}
+		// Update the list of enabled custom resources.
```

Comment:
This is creating a new list and not updating it, right?

Reply:
Yep. I've used "update" here in the sense that this is the updated list of enabled resources from the previous one (before the latest informer event propagation for "add" or "delete" took place). PLMK if this looks off or doesn't contribute much to the meaning and I'll change it.